contracts/MultiRewards.sol

```diff
@@ -447,84 +490,128 @@ contract MultiRewards is ReentrancyGuard, Pausable {
-        _totalSupply = _totalSupply.add(amount);
-        _balances[msg.sender] = _balances[msg.sender].add(amount);
-        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
-        emit Staked(msg.sender, amount);
+    function unstake(uint256 amount) external {
+        _unstake(amount, msg.sender);
     }
 
-    function withdraw(uint256 amount) public nonReentrant updateReward(msg.sender) {
-        require(amount > 0, "Cannot withdraw 0");
-        _totalSupply = _totalSupply.sub(amount);
-        _balances[msg.sender] = _balances[msg.sender].sub(amount);
-        stakingToken.safeTransfer(msg.sender, amount);
-        emit Withdrawn(msg.sender, amount);
+    function _unstake(uint256 amount, address onBehalfOf) internal {
+        UserData storage userInfo = userData[onBehalfOf];
+        if (userInfo.tokenAmount < amount)
+            revert InvalidAmount();
+        _updateReward();
+        for (uint i; i < rewardTokens.length; i ++) {
+            _calculateClaimable(onBehalfOf, rewardTokens[i]);
+        }
+        IERC20(stakingToken).safeTransfer(onBehalfOf, amount);
+
+        userInfo.tokenAmount -= amount;
+        totalStakes -= amount;
+
+        emit Unstake(onBehalfOf, amount);
     }
 
-    function getReward() public nonReentrant updateReward(msg.sender) {
+    function getReward(address _onBehalfOf, address[] memory _rewardTokens) external {
+        _getReward(_onBehalfOf, _rewardTokens);
+    }
 
-        for (uint i; i < rewardTokens.length; i++) {
-            address _rewardsToken = rewardTokens[i];
-            uint256 reward = rewards[msg.sender][_rewardsToken];
-            if (reward > 0) {
-                rewards[msg.sender][_rewardsToken] = 0;
-                IERC20(_rewardsToken).safeTransfer(msg.sender, reward);
-                emit RewardPaid(msg.sender, _rewardsToken, reward);
+    function _getReward(
+        address _user,
+        address[] memory _rewardTokens
+    ) internal whenNotPaused {
+        for (uint256 i; i < _rewardTokens.length; i ++) {
+            address token = _rewardTokens[i];
+            RewardData storage r = rewardData[token];
+            _updateReward();
+            _calculateClaimable(_user, token);
+            if (claimable[token][_user] > 0) {
+                IERC20(token).safeTransfer(_user, claimable[token][_user]);
+                r.amount -= claimable[token][_user];
+                claimable[token][_user] = 0;
+                emit RewardPaid(_user, token, claimable[token][_user]);
             }
+        }
     }
+
+    function getAllRewards() external {
+        _getReward(msg.sender, rewardTokens);
     }
 
-    function exit() external {
-        withdraw(_balances[msg.sender]);
-        getReward();
+    function updateReward() external {
+        _updateReward();
     }
 
-    /* ========== RESTRICTED FUNCTIONS ========== */
 
-    function notifyRewardAmount(address _rewardsToken, uint256 reward) external updateReward(address(0)) {
```

Comment:
Curve's staker contract doesn't appear to be drip/emission based as it never considers the rewardToken balance of its self, unlike the Gamma one which does in the `_updateReward` function, rather rewards have to be added by the `rewardsDistributor` using this `notifyRewardAmount` function.

Reply:
yes, source of rewards is different. In our case, vault is the source, so we are doing the same as Gamma here